coder/coder
# feat: Refactor CLI config-ssh to improve UX

This PR improves the UX of `coder config-ssh` with multiple QoL improvements. Replaces #1848.

- Magic block is replaced by Include statement
- Writes are only done on changes
- Inform user of changes via prompt
- Allow displaying changes via `--diff`
- Remove magic block if present
- Safer config writing via tmp-file + rename
- Parse previous `config-ssh` options, compare to new options and ask to use new (otherwise old ones are used)
- Tests for the new functionality

Fixes #1326

TODO:

- [ ] Auto-update `~/.ssh/coder` on `coder create` and `coder delete`
  - **This will be be left to a follow-up PR, or scrapped in unwanted**

---

Examples:

New config:
```
❯ coder config-ssh
> The following changes will be made to your SSH configuration:

    * Remove old config block (START-CODER/END-CODER) from /home/maf/.ssh/config
    * Add "Include coder" to /home/maf/.ssh/config
    * Write auto-generated coder config file to /home/maf/.ssh/coder

  To see changes, run diff:

    $ coder config-ssh --diff

  Continue? (yes/no) yes

You should now be able to ssh into your workspace.
For example, try running:

	$ ssh coder.work
```

Options changed:
```
❯ coder config-ssh -o ForwardAgent=yes
> New options differ from previous options:

  New options:
    * ssh-option: ForwardAgent=yes

  Previous options: none

  Use new options? (yes/no) yes

> The following changes will be made to your SSH configuration:

    * Update auto-generated coder config file in /home/maf/.ssh/coder

  To see changes, run diff:

    $ coder config-ssh --ssh-option "ForwardAgent=yes" --diff

  Continue? (yes/no) yes

You should now be able to ssh into your workspace.
For example, try running:

	$ ssh coder.work
```

Options would change => select no (no changes):
```
❯ coder config-ssh -o ForwardAgent=yes -o IdentityAgent=none
> New options differ from previous options:

  New options:
    * ssh-option: ForwardAgent=yes
    * ssh-option: IdentityAgent=none

  Previous options:
    * ssh-option: ForwardAgent=yes

  Use new options? (yes/no) no

You should now be able to ssh into your workspace.
For example, try running:

	$ ssh coder.work
```

Diff:
```
❯ coder config-ssh --diff
Changes:

  * Remove old config block (START-CODER/END-CODER) from /home/maf/.ssh/config
  * Add "Include coder" to /home/maf/.ssh/config
  * Write auto-generated coder config file to /home/maf/.ssh/coder

--- /home/maf/.ssh/config
+++ /home/maf/.ssh/config.new
@@ -1,23 +1,6 @@
+Include coder
+
 Host github.com
 	HostName github.com
 	User git

-# ------------START-CODER-----------
-# This was generated by "coder config-ssh".
-#
-# To remove this blob, run:
-#
-#    coder config-ssh --remove
-#
-# You should not hand-edit this section, unless you are deleting it.
-
-Host coder.work
-	ForwardAgent=yes
-	HostName coder.work
-	ConnectTimeout=0
-	StrictHostKeyChecking=no
-	UserKnownHostsFile=/dev/null
-	LogLevel ERROR
-	ProxyCommand "coder" --global-config "/Users/maf/Library/Application Support/coderv2" ssh --stdio work
-
-# ------------END-CODER------------

--- /home/maf/.ssh/coder
+++ /home/maf/.ssh/coder.new
@@ -0,0 +1,14 @@
+# This file is managed by coder. DO NOT EDIT.
+#
+# You should not hand-edit this file, all changes will be lost upon workspace
+# creation, deletion or when running "coder config-ssh".
+#
+# Last config-ssh options:
+#
+Host coder.work
+	HostName coder.work
+	ConnectTimeout=0
+	StrictHostKeyChecking=no
+	UserKnownHostsFile=/dev/null
+	LogLevel ERROR
+	ProxyCommand "coder" --global-config "/home/maf/.config/coderv2" ssh --stdio work
exit status 1
```
